signatures.md

```diff
@@ -52,163 +49,359 @@ Note: `Message` is used here as the object being signed, but this interface shou
+  		  //   addr - the Filecoin `Address` of the signer
+        //   sk - a private key which cryptographically links `M` to `sig`
         //
-        Sign(m Message, sk PrivateKey) (sig SignatureBytes, err error)
-
-        // Verify validates the statement: only `M` could have generated `sig`
+        Sign(msg Message, addr Address, sk PrivateKey) (sig SignatureBytes, err error)
+        
+        // Aggregate generates a BLS signature aggregate from a set of BLS 
+        // signatures. Associated `Address` must be passed to this function
+  		  // so that the aggregate can be verified. This function will pair a `sig` 
+  			// with an `Address` based on index in `sigs` and `addresses`.
+        //
+        // Out:
+        // 	 sig - a series of bytes representing a single BLS signature
+        //   err - a standard error message indicating any process issues
+        // In:
+     		//   sigs - an array of SignatureBytes {sig_1, sig_2..., sig_n}
+    		//   addrs - an array of Addresses {address_sig_1, address_sig_2..., address_sig_n}
+  			//
+  			Aggregate(sigs []SignatureBytes, addrs []Address)(sig SignatureBytes, err error)
+  
+  			// Verify validates the statement: only `M` could have generated `sig`
         // given the validator has a message `m`, a signature `sig`, and a
         // public key `pk`.
         //
         // Out:
-        //    valid - a boolean value indicating the signature is valid
-        //    err - a standard error message indicating any process issues
+        //   valid - a boolean value indicating the signature is valid
+        //   err - a standard error message indicating any process issues
+        // In:
+  			//   smsg - a SignedMessage which contains {Message, Address, SignatureBytes}
+        //   pk - a Public Key of the appropriate type
+        //
+        Verify(smsg SignedMessgage, pk PublicKey) (valid bool, err error)
+  
+  			// VerifyAggregate validates a collection of statements that are represeted by 
+  			// a signature aggregate: "For each M_i and PK_i in smsgs, only this collection 
+  			// of {M}, {PK}, and signers could have generated `smsgs.SignatureBytes`
+  			//
+        // Out:
+        //   valid - a boolean value indicating the signature is valid
+        //   err - a standard error message indicating any process issues
         // In:
-        //    m - a series of bytes representing the signed message
-        //    pk - the public key belonging to the signer `M`
-        //    sig - a series of bytes representing a signature usually `r`|`s`
+        //   smsgs - a SignedMessage which contains {[]Message, []Address, SignatureBytes}
+        //   pks - an array of Public Keys of the appropriate type associated with a message
+  			// 	 at index `i` in []Message.
         //
-        Verify(m Messgage, pk PublicKey, sig SignatureBytes) (valid bool, err error)
+  			VerifyAggregate(smsgs SignedMessgage,  pks []PublicKeys) (valid bool, err error)
```

Comment:
are there multiple signed messages here? also does the signature get passed in? or how does this work?

Reply:
Updating name "smgs" --> "smsg" to reduce confusion. Any `SignedMessage` containing a signature aggregate will contain the signature aggregate as well as all messages contained therein. So we only need the one `SignedMessage` as input.

Setting aside golang syntax, suppose there is `X = SignedMessage`. To verify a signature aggregate, VerifyAggregate will perform something like: 

- For all messages `m_i` in `X.Msgs`, `h_i = hash2curve(m_i)`
- Compute `S = deserialize(X.SignatureBytes)`
- Compute `T = e(h_1, pk_i)*...*e(h_n, pk_n)`
- If `T = e(S, P2)` => valid